ctapipe/io/toymodel.py

```diff
@@ -5,95 +5,132 @@
 import logging
 
 import numpy as np
-from ctapipe.image import toymodel
-from scipy.stats import norm
 import astropy.units as u
 
-from .containers import DataContainer
+from ..containers import (
+    DataContainer, DL1CameraContainer, EventIndexContainer,
+)
+from ..core import traits
+from ..core import TelescopeComponent
+from ..image import toymodel
+from .eventsource import EventSource
 
 logger = logging.getLogger(__name__)
 
 
-def toymodel_event_source(geoms, max_events=100, single_tel=False, n_channels=1,
-                          n_samples=25, p_trigger=0.3):
-    """
-    An event source that produces array
-    Parameters
-    ----------
-    geoms : list of CameraGeometry instances
-        Geometries for the telescopes to simulate
-    max_events : int, default: 100
-        maximum number of events to create
-    n_channels : int
-        how many channels per telescope
-    n_samples : int
-        how many adc samples per pixel
-    p_trigger : float
-        mean trigger probability for the telescopes
-    """
-    n_telescopes = len(geoms)
-    container = DataContainer()
-    container.meta['toymodel__max_events'] = max_events
-    container.meta['source'] = "toymodel"
-    tel_ids = np.arange(n_telescopes)
-
-    for event_id in range(max_events):
-
-        n_triggered = np.random.poisson(n_telescopes * 0.3)
-        if n_triggered > n_telescopes:
-            n_triggered = n_telescopes
-
-        triggered_tels = np.random.choice(tel_ids, n_triggered, replace=False)
-
-        container.r0.event_id = event_id
-        container.r0.tels_with_data = triggered_tels
-        container.count = event_id
-
-        # handle single-telescope case (ignore others:
-        if single_tel:
-            if single_tel not in container.r0.tels_with_data:
+class ToyEventSource(EventSource, TelescopeComponent):
+
+    trigger_probability = traits.FloatTelescopeParameter(
+        default_value=0.5, help='Probability that the telescope has an event',
+    )
```

Comment:
I think all of these need the usual `.tag(config=True)` for them to be config parameters that work properly.

Reply:
Ah, of course. That's really annoying

Reply:
yes, it's one of the reasons I want to get rid of that config system. I had a hand-made version that does a large fraction of the functionality with only traitlets, and not traitlets.config, but there are still some nice things in there that are hard to reproduce.